File: surface/widgets/swiftui/components/widgets/concepts/process-human/SlaTimer.swift
import SwiftUI

enum SlaTimerWidgetState: String {
    case onTrack, warning, critical, breached, paused
}

enum SlaTimerEvent {
    case tick, warningThreshold, pause, criticalThreshold, breach, resume
}

func slaTimerReduce(state: SlaTimerWidgetState, event: SlaTimerEvent) -> SlaTimerWidgetState {
    switch state {
    case .onTrack:
        if event == .warningThreshold { return .warning }
        if event == .pause { return .paused }
        return state
    case .warning:
        if event == .criticalThreshold { return .critical }
        if event == .pause { return .paused }
        return state
    case .critical:
        if event == .breach { return .breached }
        if event == .pause { return .paused }
        return state
    case .breached:
        return state
    case .paused:
        if event == .resume { return .onTrack }
        return state
    }
}

struct SlaTimerView: View {
    var dueAt: String
    var status: String
    var warningThreshold: Double = 0.7
    var criticalThreshold: Double = 0.9
    var showElapsed: Bool = true
    var startedAt: String? = nil
    var onBreach: (() -> Void)? = nil
    var onWarning: (() -> Void)? = nil
    var onCritical: (() -> Void)? = nil

    @State private var widgetState: SlaTimerWidgetState = .onTrack
    @State private var remaining: TimeInterval = 0
    @State private var elapsed: TimeInterval = 0
    @State private var progress: Double = 0
    @State private var breachedFired: Bool = false
    @State private var warningFired: Bool = false
    @State private var criticalFired: Bool = false

    private let timer = Timer.publish(every: 1, on: .main, in: .common).autoconnect()

    private let phaseLabels: [SlaTimerWidgetState: String] = [
        .onTrack: "On Track",
        .warning: "Warning",
        .critical: "Critical",
        .breached: "Breached",
        .paused: "Paused"
    ]

    private var phaseColor: Color {
        switch widgetState {
        case .onTrack: return .green
        case .warning: return .orange
        case .critical: return .red
        case .breached: return .red
        case .paused: return .gray
        }
    }

    private func formatCountdown(_ ms: TimeInterval) -> String {
        guard ms > 0 else { return "00:00:00" }
        let totalSeconds = Int(ms / 1000)
        let hours = totalSeconds / 3600
        let minutes = (totalSeconds % 3600) / 60
        let seconds = totalSeconds % 60
        return String(format: "%02d:%02d:%02d", hours, minutes, seconds)
    }

    private func formatElapsedTime(_ ms: TimeInterval) -> String {
        guard ms > 0 else { return "0s" }
        let totalSeconds = Int(ms / 1000)
        let hours = totalSeconds / 3600
        let minutes = (totalSeconds % 3600) / 60
        let seconds = totalSeconds % 60
        if hours > 0 { return "\(hours)h \(minutes)m \(seconds)s" }
        if minutes > 0 { return "\(minutes)m \(seconds)s" }
        return "\(seconds)s"
    }

    private var dueTime: Date {
        ISO8601DateFormatter().date(from: dueAt) ?? Date()
    }

    private var startTime: Date {
        if let started = startedAt {
            return ISO8601DateFormatter().date(from: started) ?? Date()
        }
        return Date()
    }

    private var totalDuration: TimeInterval {
        dueTime.timeIntervalSince(startTime) * 1000
    }

    private func tick() {
        let now = Date()
        let rem = max(0, dueTime.timeIntervalSince(now) * 1000)
        let elap = now.timeIntervalSince(startTime) * 1000
        let prog = totalDuration > 0 ? min(1, elap / totalDuration) : 1

        remaining = rem
        elapsed = elap
        progress = prog

        if rem <= 0 && !breachedFired {
            breachedFired = true
            widgetState = slaTimerReduce(state: widgetState, event: .breach)
            onBreach?()
        } else if prog >= criticalThreshold && !criticalFired && rem > 0 {
            criticalFired = true
            widgetState = slaTimerReduce(state: widgetState, event: .criticalThreshold)
            onCritical?()
        } else if prog >= warningThreshold && !warningFired && rem > 0 {
            warningFired = true
            widgetState = slaTimerReduce(state: widgetState, event: .warningThreshold)
            onWarning?()
        }
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            // Countdown display
            Text(widgetState == .breached ? "BREACHED" : formatCountdown(remaining))
                .font(.system(.title, design: .monospaced))
                .fontWeight(.bold)
                .foregroundColor(phaseColor)
                .accessibilityLabel("Time remaining: \(formatCountdown(remaining))")

            // Phase label
            Text(phaseLabels[widgetState] ?? "")
                .font(.subheadline)
                .fontWeight(.medium)
                .foregroundColor(phaseColor)

            // Progress bar
            GeometryReader { geo in
                ZStack(alignment: .leading) {
                    RoundedRectangle(cornerRadius: 4)
                        .fill(Color.gray.opacity(0.2))
                        .frame(height: 8)
                    RoundedRectangle(cornerRadius: 4)
                        .fill(phaseColor)
                        .frame(width: geo.size.width * CGFloat(progress), height: 8)
                        .animation(.easeInOut(duration: 0.3), value: progress)
                }
            }
            .frame(height: 8)
            .accessibilityLabel("SLA progress: \(Int(progress * 100))%")
            .accessibilityValue("\(Int(progress * 100)) percent")

            // Elapsed time
            if showElapsed {
                Text("Elapsed: \(formatElapsedTime(elapsed))")
                    .font(.caption)
                    .foregroundColor(.secondary)
                    .accessibilityLabel("Elapsed time: \(formatElapsedTime(elapsed))")
            }

            // Pause/Resume
            if widgetState != .breached {
                Button(widgetState == .paused ? "Resume" : "Pause") {
                    if widgetState == .paused {
                        widgetState = slaTimerReduce(state: widgetState, event: .resume)
                    } else {
                        widgetState = slaTimerReduce(state: widgetState, event: .pause)
                    }
                }
                .accessibilityLabel(widgetState == .paused ? "Resume timer" : "Pause timer")
            }
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel("SLA timer: \(phaseLabels[widgetState] ?? "")")
        .accessibilityRole(.timer)
        .onAppear { tick() }
        .onReceive(timer) { _ in
            guard widgetState != .paused else { return }
            tick()
        }
    }
}

File: surface/widgets/compose/components/widgets/concepts/process-human/SlaTimer.kt
package com.clef.surface.widgets.concepts.processhuman

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.delay

// --- State machine ---

enum class SlaTimerState { OnTrack, Warning, Critical, Breached, Paused }

sealed class SlaTimerEvent {
    object Tick : SlaTimerEvent()
    object WarningThreshold : SlaTimerEvent()
    object Pause : SlaTimerEvent()
    object CriticalThreshold : SlaTimerEvent()
    object Breach : SlaTimerEvent()
    object Resume : SlaTimerEvent()
}

fun slaTimerReduce(state: SlaTimerState, event: SlaTimerEvent): SlaTimerState = when (state) {
    SlaTimerState.OnTrack -> when (event) {
        is SlaTimerEvent.Tick -> SlaTimerState.OnTrack
        is SlaTimerEvent.WarningThreshold -> SlaTimerState.Warning
        is SlaTimerEvent.Pause -> SlaTimerState.Paused
        else -> state
    }
    SlaTimerState.Warning -> when (event) {
        is SlaTimerEvent.Tick -> SlaTimerState.Warning
        is SlaTimerEvent.CriticalThreshold -> SlaTimerState.Critical
        is SlaTimerEvent.Pause -> SlaTimerState.Paused
        else -> state
    }
    SlaTimerState.Critical -> when (event) {
        is SlaTimerEvent.Tick -> SlaTimerState.Critical
        is SlaTimerEvent.Breach -> SlaTimerState.Breached
        is SlaTimerEvent.Pause -> SlaTimerState.Paused
        else -> state
    }
    SlaTimerState.Breached -> when (event) {
        is SlaTimerEvent.Tick -> SlaTimerState.Breached
        else -> state
    }
    SlaTimerState.Paused -> when (event) {
        is SlaTimerEvent.Resume -> SlaTimerState.OnTrack
        else -> state
    }
}

// --- Helpers ---

private fun formatCountdown(ms: Long): String {
    if (ms <= 0) return "00:00:00"
    val totalSeconds = (ms / 1000).toInt()
    val h = totalSeconds / 3600
    val m = (totalSeconds % 3600) / 60
    val s = totalSeconds % 60
    return "${"%02d".format(h)}:${"%02d".format(m)}:${"%02d".format(s)}"
}

private fun formatElapsed(ms: Long): String {
    if (ms <= 0) return "0s"
    val totalSeconds = (ms / 1000).toInt()
    val h = totalSeconds / 3600
    val m = (totalSeconds % 3600) / 60
    val s = totalSeconds % 60
    return when {
        h > 0 -> "${h}h ${m}m ${s}s"
        m > 0 -> "${m}m ${s}s"
        else -> "${s}s"
    }
}

private val PHASE_LABELS = mapOf(
    SlaTimerState.OnTrack to "On Track",
    SlaTimerState.Warning to "Warning",
    SlaTimerState.Critical to "Critical",
    SlaTimerState.Breached to "Breached",
    SlaTimerState.Paused to "Paused"
)

private val PHASE_COLORS = mapOf(
    SlaTimerState.OnTrack to Color(0xFF16A34A),
    SlaTimerState.Warning to Color(0xFFCA8A04),
    SlaTimerState.Critical to Color(0xFFEA580C),
    SlaTimerState.Breached to Color(0xFFDC2626),
    SlaTimerState.Paused to Color(0xFF6B7280)
)

@Composable
fun SlaTimer(
    dueAt: String,
    status: String,
    modifier: Modifier = Modifier,
    warningThreshold: Float = 0.7f,
    criticalThreshold: Float = 0.9f,
    showElapsed: Boolean = true,
    startedAt: String? = null,
    onBreach: () -> Unit = {},
    onWarning: () -> Unit = {},
    onCritical: () -> Unit = {}
) {
    var state by remember { mutableStateOf(SlaTimerState.OnTrack) }
    var remaining by remember { mutableLongStateOf(0L) }
    var elapsed by remember { mutableLongStateOf(0L) }
    var progress by remember { mutableFloatStateOf(0f) }
    val breached = remember { mutableStateOf(false) }
    val warningFired = remember { mutableStateOf(false) }
    val criticalFired = remember { mutableStateOf(false) }

    val dueTime = remember(dueAt) {
        try { java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss").parse(dueAt)?.time ?: System.currentTimeMillis() }
        catch (_: Exception) { System.currentTimeMillis() }
    }
    val startTime = remember(startedAt) {
        if (startedAt != null) try { java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss").parse(startedAt)?.time ?: System.currentTimeMillis() }
        catch (_: Exception) { System.currentTimeMillis() }
        else System.currentTimeMillis()
    }
    val totalDuration = dueTime - startTime

    // Tick effect
    LaunchedEffect(state, dueTime, startTime) {
        if (state == SlaTimerState.Paused) return@LaunchedEffect
        while (true) {
            val now = System.currentTimeMillis()
            val rem = maxOf(0L, dueTime - now)
            val elap = now - startTime
            val prog = if (totalDuration > 0) minOf(1f, elap.toFloat() / totalDuration) else 1f

            remaining = rem
            elapsed = elap
            progress = prog

            state = slaTimerReduce(state, SlaTimerEvent.Tick)

            if (rem <= 0L && !breached.value) {
                breached.value = true
                state = slaTimerReduce(state, SlaTimerEvent.Breach)
                onBreach()
            } else if (prog >= criticalThreshold && !criticalFired.value && rem > 0) {
                criticalFired.value = true
                state = slaTimerReduce(state, SlaTimerEvent.CriticalThreshold)
                onCritical()
            } else if (prog >= warningThreshold && !warningFired.value && rem > 0) {
                warningFired.value = true
                state = slaTimerReduce(state, SlaTimerEvent.WarningThreshold)
                onWarning()
            }

            delay(1000)
        }
    }

    val phaseColor = PHASE_COLORS[state] ?: Color.Gray
    val progressPercent = (progress * 100).toInt()

    Column(modifier = modifier.semantics { contentDescription = "SLA timer: ${PHASE_LABELS[state]}" }) {
        // Countdown
        Text(
            if (state == SlaTimerState.Breached) "BREACHED" else formatCountdown(remaining),
            fontWeight = FontWeight.Bold,
            fontSize = 20.sp,
            color = phaseColor
        )

        // Phase label
        Text(
            PHASE_LABELS[state] ?: "",
            fontSize = 13.sp,
            color = phaseColor,
            modifier = Modifier.padding(top = 4.dp)
        )

        // Progress bar
        LinearProgressIndicator(
            progress = { progress },
            modifier = Modifier.fillMaxWidth().height(8.dp).padding(top = 8.dp),
            color = phaseColor
        )

        // Elapsed time
        if (showElapsed) {
            Text(
                "Elapsed: ${formatElapsed(elapsed)}",
                fontSize = 12.sp,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.padding(top = 4.dp)
            )
        }

        // Pause/Resume
        if (state != SlaTimerState.Breached) {
            TextButton(
                onClick = {
                    state = if (state == SlaTimerState.Paused) slaTimerReduce(state, SlaTimerEvent.Resume)
                    else slaTimerReduce(state, SlaTimerEvent.Pause)
                },
                modifier = Modifier.padding(top = 4.dp)
            ) {
                Text(if (state == SlaTimerState.Paused) "Resume" else "Pause", fontSize = 13.sp)
            }
        }
    }
}
